Content.Server/Fluids/EntitySystems/MoppingSystem.cs

```diff
@@ -0,0 +1,350 @@
+                        msg = "mopping-system-hand-squeeze-dry";
+                    }
+
+                }
+                else
+                {
+                    transferAmount = FixedPoint2.Min(refillableSolution.AvailableVolume, currentVolume); //Transfer all liquid from the tool to the container, but only if it will fit.
+                    msg = "mopping-system-refillable-success";
+                    delay = 1.0f;
+                }
+
+                donor = used; // the mop/tool Uid
+                donorSolutionName = "absorbed"; // by definition on AbsorbentComponent
+
+                acceptor  = target; // the refillable container's Uid
+                acceptorSolutionName = refillable.Solution;
+
+                // Set delay/popup/sound if nondefault. Popup and sound will only play on a successful doAfter.
+
+                sfx = component.TransferSound;
+            }
+        }
+        else if (TryComp<DrainableSolutionComponent>(target, out var drainable) // We can take solution from the target
+                && currentVolume <= 0 ) // tool is dry
+        {
+            // These return conditions will abort BEFORE the do_after is called:
+            if (!_solutionSystem.TryGetDrainableSolution(target, out var drainableSolution))
+            {
+                return;
+            }
+            else if (drainableSolution.CurrentVolume <= 0) // target container is empty (liquid source)
+            {
+                msg = "mopping-system-target-container-empty";
+                user.PopupMessage(user, Loc.GetString(msg, ("target", target))); // play message now because we are returning.
+                return;
+            }
+            else
+            {
+                // Determine transferAmount
+                transferAmount = FixedPoint2.Min(availableVolume * 0.5, drainableSolution.CurrentVolume); // Let's transfer up to to half the tool's available capacity to the tool.
+
+                donor = target; // the drainable container's Uid
+                donorSolutionName = drainable.Solution;
+
+                acceptor  = used; // the mop/tool Uid
+                acceptorSolutionName = "absorbed"; // by definition on AbsorbentComponent
+
+                // Set delay/popup/sound if nondefault. Popup and sound will only play on a successful doAfter.
+                // default delay is fine for this case.
+                msg = "mopping-system-drainable-success";
+                sfx = component.TransferSound;
+            }
+        }
+        else return;
+
+        var doAfterArgs = new DoAfterEventArgs(user, delay, target: target)
+        {
+            BreakOnUserMove = true,
+            BreakOnStun = true,
+            BreakOnDamage = true,
+            MovementThreshold = 0.2f,
+            BroadcastCancelledEvent = new TransferCancelledEvent()
+            {
+                Target = target,
+                Component = component // (the AbsorbentComponent)
```

Comment:
Your code assumes that the components still exist *sometime* in the future when the complete/cancelled handlers are called. Transferring liquids to a deleted component is bad (C++ would literally crash your program for doing this in the best case). At least check the `Deleted` flag on the components in the handlers (weak_ptr components when?), or send EUIDs instead.

Reply:
On closer inspection, I don't actually use the component anywhere for the cancelled case. So I simply removed it from `BroadcastCancelledEvent`... does this resolve that problem?

Reply:
Never mind, I totally do. Ok, I'll have a look at checking for `Deleted` flags.

Reply:
by EUID do you just mean Entity UID?

Reply:
I think this handles it, unless I missed your point.

![image](https://user-images.githubusercontent.com/97707302/156909449-83b2c7f5-bbff-4413-90b6-edc149826653.png)